### PR TITLE
pkg/bpf: Protect each uintptr with runtime.KeepAlive

### PR DESCRIPTION
**Pull request by brb**
The Go's GC is unable to track objects referred by `uintptr` \[1\] \[2\].

A consequence of this is that the GC might collect such objects when a blocking syscall is executed. This might corrupt memory of other Go objects e.g. if the reclaimed memory is used for new objects and the kernel modifies the objects in the context of the blocking syscall.

To prevent from this, we protect each pointer to objects with `runtime.KeepAlive()` which marks the objects as alive at least until `runtime.KeepAlive()` has been called.

**FOR REVIEWERS**:
- Please double-check (I might have missed, and I don't see any SA tool which can check this) that each reference to objects which references are converted to `uintptr` and then passed to `Syscall*()` is protected with `runtime.KeepAlive()`.
- For v1.6 and v1.5 I need to create an additional backport to fix `perf_linux.go` which in v1.7 was replaced by `cilium/ebpf`. I will do it once this PR gets backported.

\[1\]: https://github.com/golang/go/issues/13372#issuecomment-160655731
\[2\]: https://utcc.utoronto.ca/~cks/space/blog/programming/GoUintptrVsUnsafePointer

Developed together with @aanm and @borkmann.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10168)

### CODE REVIEW

**Comment by brb** [comment]
test-me-please

**Comment by aanm** [comment]
```
"github.com/cilium/cilium/pkg/datapath/loader.DatapathSHA="'  -o maptool
# github.com/cilium/cilium/pkg/bpf
../../pkg/bpf/map_linux.go:610:2: undefined: runtime
../../pkg/bpf/map_linux.go:611:2: undefined: runtime
../../pkg/bpf/map_linux.go:612:2: undefined: runtime
../../pkg/bpf/map_linux.go:613:2: undefined: runtime
../../pkg/bpf/map_linux.go:614:2: undefined: runtime
../../pkg/bpf/map_linux.go:751:2: undefined: runtime
../../pkg/bpf/map_linux.go:752:2: undefined: runtime
../../pkg/bpf/map_linux.go:753:2: undefined: runtime
../../pkg/bpf/map_linux.go:754:2: undefined: runtime
make[1]: *** [maptool] Error 2
make[1]: Leaving directory `/home/travis/gopath/src/github.com/cilium/cilium/tools/maptool'
make: *** [unit-tests] Error 2
```

**Comment by florianl** [comment]
Maybe just an idea:
By default all bpf calls are using `runtime.KeepAlive()` in [cilium/ebpf](https://github.com/cilium/ebpf/blob/master/internal/syscall.go#L13). So maybe, this can be used and ` pkg/bpf/` wraps this and adds on top only the cilium specific parts, like `BPFSyscallDurationEnabled`.

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
@florianl This is just a temporary fix. We are planning to migrate to cilium/ebpf soonish.

**Comment by joestringer** [comment]
Travis failed:

```
# github.com/cilium/cilium/pkg/bpf
pkg/bpf/map_linux.go:611:2: unreachable code
```

**Comment by aanm** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
PTAL (please see a note for reviewers in the PR description).

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
I did a close review to this PR, I've pushed force after double checking with Martynas about 2 places where we were unnecessary having `runtime.KeepAlive()`

File: pkg/bpf/bpf_linux.go [hunk 1]
```diff
@@ -134,22 +137,27 @@ func UpdateElement(fd int, key, value unsafe.Pointer, flags uint64) error {
 		flags: uint64(flags),
 	}
 
-	return UpdateElementFromPointers(fd, uintptr(unsafe.Pointer(&uba)), unsafe.Sizeof(uba))
+	ret := UpdateElementFromPointers(fd, unsafe.Pointer(&uba), unsafe.Sizeof(uba))
+	runtime.KeepAlive(key)
+	runtime.KeepAlive(value)
+	runtime.KeepAlive(&uba)
```
Comment (joestringer) [inline pkg/bpf/bpf_linux.go]: This case is using `runtime.KeepAlive()` to keep a pointer to a stack value alive, yet this pointer is passed as an `unsafe.Pointer` to `UpdateElementFromPointers` which I would expect to retain the reference-counting semantics at least for the duration of this function's usage of the pointer. The callee, `UpdateElementFromPointers` is already handling the `KeepAlive()` of this pointer in its own implementation, so I don't think it's necessary here?

File: pkg/bpf/map_linux.go [hunk 2]
```diff
@@ -601,11 +602,19 @@ func (m *Map) DumpWithCallback(cb DumpCallback) error {
...
 			}
 			return err
 		}
 	}
+
+	runtime.KeepAlive(&key)
+	runtime.KeepAlive(&nextKey)
+	runtime.KeepAlive(&bpfCurrentKey)
+	runtime.KeepAlive(&value)
+	runtime.KeepAlive(&bpfNextKey)
```
Comment (joestringer) [inline pkg/bpf/map_linux.go]: Given that this function applies `uintptr` casting for `key`, `nextKey`, `value`, I understand the `KeepAlive()` of those. I would have thought that if this function changes the `bpfCurrentKey` and `bpfNextKey` to `unsafe.Pointer` then these two `KeepAlive()`s should not be necessary; the lower-layer functions will apply that for the duration of the actual syscalls anyway.

Comment (joestringer) [reply pkg/bpf/map_linux.go]: Then if the iteration may or may not cause subsequent references to these variables depending on runtime information, presumably it's not safe for the compiler to mark the values as unused / ready to GC until this function returns anyway ...?

Comment (joestringer) [reply pkg/bpf/map_linux.go]: (If this leads to any change in approach then it would also apply to the other dump function)

File: pkg/bpf/map_linux.go [hunk 3]
```diff
@@ -940,6 +955,10 @@ func (m *Map) DeleteAll() error {
 			return err
 		}
 	}
+
+	runtime.KeepAlive(&nextKey)
```
Comment (joestringer) [inline pkg/bpf/map_linux.go]: Is this case necessary? It doesn't follow the same basic `uintptr` pattern; `GetFirstKey()` should presumably provide the protection necessary here.

**Review by joestringer — CHANGES_REQUESTED** [review]
I have a few queries below, I think the base format is that wherever we cast to `uintptr`, we need a subsequent `runtime.KeepAlive()` following the usage in a syscall. That makes sense. There's a couple of cases below where the changes are beyond this specific pattern, and a couple of cases where `runtime.KeepAlive()` is applied twice for the same pointer.

I ran this:
```
$ git grep -l uintptr -- ':^vendor/'
pkg/bpf/bpf_linux.go
pkg/bpf/map_linux.go
pkg/bpf/prog_linux.go
pkg/endpoint/connector/ipvlan.go
pkg/fqdn/cache.go
pkg/fqdn/dnsproxy/udp.go
pkg/maps/cidrmap/cidrmap.go
pkg/proxy/dialer.go
pkg/proxy/socket.go
```

Upon closer inspection, most of it is handling an `fd` or actually doing arithmetic. Looks like just the ipvlan case needs this treatment to fix it up. 

Otherwise LGTM, nice sleuthing work on this!

File: pkg/bpf/map_linux.go (hunk 3 above)
Comment (aanm) [reply pkg/bpf/map_linux.go]: this one is keeping the reference for `nextKey`, the `DeleteElement` and `GetFirstKey` are keeping the reference of `nextKey[0]`

@brb maybe it's better to add a comment for this case.

**Review by aanm — APPROVED** [review]

File: pkg/bpf/bpf_linux.go (hunk 1 above)
Comment (brb) [reply pkg/bpf/bpf_linux.go]: > to keep a pointer to a stack value alive

OT: in Go, values on stack might escape to heap.

> The callee, UpdateElementFromPointers is already handling the KeepAlive() of this pointer in its own implementation, so I don't think it's necessary here?

That's correct. I've introduced a bit of cargo cult programming:S

File: pkg/bpf/map_linux.go (hunk 2 above)
Comment (brb) [reply pkg/bpf/map_linux.go]: > I would have thought that if this function changes the bpfCurrentKey and bpfNextKey to unsafe.Pointer then these two KeepAlive()s should not be necessary

Re `bpfNextKey` and `bpfCurrentKey`: that's correct (a leftover from changing the function annotations).

> it's not safe for the compiler to mark the values as unused / ready to GC until this function returns anyway ...?

Just validated that it's indeed the case, and we can remove `runtime.KeepAlive()` from here and other dump functions.

File: pkg/bpf/map_linux.go (hunk 3 above)
Comment (brb) [reply pkg/bpf/map_linux.go]: This can be eliminated, as `dumpParser()` in the loop will protect `nextKey`.

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
Thanks for the review, PTAL.

**Review by aanm — APPROVED** [review]

File: pkg/bpf/map_linux.go [hunk 4]
```diff
@@ -940,6 +943,8 @@ func (m *Map) DeleteAll() error {
 			return err
 		}
 	}
+
+	return nil
```
Comment (borkmann) [inline pkg/bpf/map_linux.go]: Why is this change needed?

File: pkg/bpf/map_linux.go [hunk 5]
```diff
@@ -601,11 +601,13 @@ func (m *Map) DumpWithCallback(cb DumpCallback) error {
...
 		if err := GetNextKeyFromPointers(m.fd, bpfCurrentKeyPtr, bpfCurrentKeySize); err != nil {
 			if err == io.EOF { // end of map, we're done iterating
-				return nil
+				break
 			}
 			return err
 		}
 	}
+
+	return nil
```
Comment (borkmann) [inline pkg/bpf/map_linux.go]: Does this need to be part of the fix? Just wondering wrt backporting.

**Review by borkmann — APPROVED** [review]

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

File: pkg/bpf/prog_linux.go [hunk 6]
```diff
@@ -122,6 +125,8 @@ func GetProgInfoByFD(fd int) (ProgInfo, error) {
 		duration = spanstat.Start()
 	}
 	ret, _, err := unix.Syscall(unix.SYS_BPF, BPF_OBJ_GET_INFO_BY_FD, uintptr(unsafe.Pointer(&attr)), unsafe.Sizeof(attr))
```
Comment (jrfastab) [inline pkg/bpf/prog_linux.go:127]: Do we also need, runtime.KeepAlive(&attr) here?

Comment (brb) [reply pkg/bpf/prog_linux.go:127]: @jrfastab Good catch, `runtime.KeepAlive(&attrInfo)` below should have been `runtime.KeepAlive(&attr)` instead. However, neither public functions from `prog_linux.go` are being used, so maybe it makes sense to backport https://github.com/cilium/cilium/pull/10187 to the `v1.{5,6,7}` instead of fixing the dead code?

Comment (jrfastab) [reply pkg/bpf/prog_linux.go:127]: I agree backporting dead-code removal makes sense here. I'll rework the backport series as well.
